tick/eval/eval_binary_node.go

```diff
@@ -0,0 +1,334 @@
+}
+
+func (e *EvalBinaryNode) evalWithNodes(scope *tick.Scope, executionState ExecutionState, left, right NodeEvaluator) (resultContainer, *ErrSide) {
+	// TODO: handle errors, for real dynamic values! for example refNode: first is float then int64
+	if e.evaluationFn == nil {
+		err := e.validateEvaluation(scope, executionState, e.operator, left, right)
+		if err != nil {
+			return boolFalseResultContainer, &ErrSide{error: err}
+		}
+	}
+
+	evaluationResult, err := e.evaluationFn(scope, executionState, left, right)
+
+	// This case can in dynamic nodes,
+	// for example: RefNode("value") > NumberNode("float64")
+	// in the first evaluation "value" is float64 so we will have float64 > float64 comparison fn
+	// after the first evaluation, let's assume that "value" is changed to int64 - we need to change
+	// the comparison fn
+	if err != nil {
+		if typeGuardErr, isTypeGuardError := err.error.(ErrTypeGuardFailed); isTypeGuardError {
+			// Fix the type info, thanks to the type guard info
+			if err.IsLeftSide {
+				e.leftSideType = typeGuardErr.ActualType
+			}
+
+			if err.IsRightSide {
+				e.rightSideType = typeGuardErr.ActualType
+			}
+
+			// re-find the evaluation fn
+			e.evaluationFn = e.findEvaluationFn(e.leftSideType, e.rightSideType)
+
+			// recurse (so we handle nil evaluationFn, and etc)
+			return e.evalWithNodes(scope, executionState, left, right)
+		}
+	}
+
+	return evaluationResult, err
+}
+
+// evaluateDynamicNode fetches the value of the right and left node at evaluation time (aka "runtime")
+// and find the matching evaluation function for the givne types - this is where the "specialisation" happens.
+func (e *EvalBinaryNode) evaluateDynamicNode(scope *tick.Scope, executionState ExecutionState, left, right NodeEvaluator) (resultContainer, *ErrSide) {
+	var leftSideType ValueType
+	var rightSideType ValueType
+	var err error
+
+	// For getting the type we must pass new execution state, since the node can be stateful (like function call)
+	// and on the second in the specialiszation we might loose the correct state
+	// For example: "count() == 1"
+	//  1. we evaluate the left side and counter is 1 (upper ^ in this function)
+	//  2. we evaluate the second time in "EvalBool"
+	typeExecutionState := CreateExecutionState()
+
+	if leftSideType, err = left.GetType(scope, typeExecutionState); err != nil {
+		return emptyResultContainer, &ErrSide{error: err, IsLeftSide: true}
+	}
+
+	if rightSideType, err = right.GetType(scope, typeExecutionState); err != nil {
+		return emptyResultContainer, &ErrSide{error: err, IsRightSide: true}
+	}
+
+	e.leftSideType = leftSideType
+	e.rightSideType = rightSideType
+
+	e.evaluationFn = e.findEvaluationFn(e.leftSideType, e.rightSideType)
+
+	return e.evalWithNodes(scope, executionState, left, right)
+}
+
+func (e *EvalBinaryNode) validateEvaluation(scope *tick.Scope, executionState ExecutionState, operator tick.TokenType, left, right NodeEvaluator) error {
```

Comment:
As far as I can tell this function always returns an error. It just figures out the most accurate error to return. Is that correct? If so can we rename it to `constructError` or `determineError` or something?